src/iop/prover.rs

```diff
@@ -0,0 +1,821 @@
+                .zip(&m_star.evals_on_domain_b.col.evals)
+                .step_by(step)
+                .map(|(row, col)| (alpha - row) * (beta - col))
+                .collect();
+            algebra::batch_inversion(&mut inverse);
+            inverse
+        };
+
+        let inverses_a = compute_denominators(&a_star);
+        let inverses_b = compute_denominators(&b_star);
+        let inverses_c = compute_denominators(&c_star);
+
+        let p_evals_on_K: Vec<_> = a_star
+            .evals_on_domain_b
+            .val_row_col
+            .evals
+            .par_iter()
+            .zip(&b_star.evals_on_domain_b.val_row_col.evals)
+            .zip(&c_star.evals_on_domain_b.val_row_col.evals)
+            .step_by(step)
+            .zip(&inverses_a)
+            .zip(&inverses_b)
+            .zip(&inverses_c)
+            .map(|(((((&a_k, b_k), c_k), a_inv), b_inv), c_inv)| {
+                eta_a * a_k * a_inv + eta_b * b_k * b_inv + eta_c * c_k * c_inv
+            })
+            .collect();
+        end_timer!(p_evals_time);
+
+        /* Compute the boundary polynomial U_2(X) for p(X) - T(alpha,beta)/|K|.
+         */
+        let u_2_time = start_timer!(|| "Compute u_2 poly");
+        let (u_2, normalized_v) = BoundaryPolynomial::from_non_coboundary_polynomial_evals(
+            EvaluationsOnDomain::from_vec_and_domain(p_evals_on_K, domain_k.clone()),
+        )
+        .map_err(|e| {
+            end_timer!(u_2_time);
+            end_timer!(round_time);
+            e
+        })?;
+
+        let u_2 = u_2.polynomial();
+        end_timer!(u_2_time);
+
+        assert!(u_2.degree() <= domain_k.size() - 1);
+
+        // Compute the shifted boundary polynomial U_2(gX)
+        let u_2_g_time = start_timer!(|| "Compute u_2_g poly");
+        let u_2_g = {
+            let mut u_2_g = u_2.clone();
+            let g_s = domain_k.elements().collect::<Vec<F>>();
+            u_2_g
+                .coeffs
+                .par_iter_mut()
+                .zip(g_s)
+                .for_each(|(a, b)| *a *= b);
+            u_2_g
+        };
+        end_timer!(u_2_g_time);
+
+        assert!(u_2_g.degree() <= domain_k.size() - 1);
+
+        /* Compute the quotient polynomial h_2(X) for the inner sumcheck identity.
+         */
+
+        let denom_eval_time = start_timer!(|| "Computing denominator evals on domain B");
+        let m_denom = |m_star: &MatrixArithmetization<F>| -> Vec<F> {
+            m_star
+                .evals_on_domain_b
+                .row
+                .evals
+                .par_iter()
+                .zip(&m_star.evals_on_domain_b.col.evals)
+                .zip(&m_star.evals_on_domain_b.row_col.evals)
+                .map(|((&r, c), r_c)| alpha * beta - beta * r - alpha * c + r_c)
+                .collect()
+        };
+        let a_denom = m_denom(a_star);
+        let b_denom = m_denom(b_star);
+        let c_denom = m_denom(c_star);
+        end_timer!(denom_eval_time);
```

Comment:
To clarify/correct my statement: we already determined `a_denom`, `b_denom`, `c_denom` over domain `K` in the above computation for the `inverses_a`, `inverses_b`, `inverses_c`. Let us optimize here as discussed offline:
- First compute `a_denom`, `b_denom`, `c_denom`  and `a_poly_over_b`, and 
- then compute the `inverses_a`, `inverses_b`, `inverses_c` over domain `K``using the `*_denom`s, and calculate the non-centered `f_poly_over_k`.

The rest is as before: we use `from_non_coboundary()` on `f_poly_over_k` to obtain the boundary poly `U_2(X)`, compute its g-shift and from it the coefficients of the `f_poly(X)`. (Note that since we need `U_2(X)` anyway, this is cheaper than calling an FFT on `f_poly_over_k`.) We compute `f_poly_over_b` via FFT, etc.,etc.

Reply:
Done. Actually did a slightly more general refactoring of the computations to further reduce duplicated computations.